client/components/data/query-products-list/index.ts

```diff
@@ -0,0 +1,57 @@
+import { useEffect } from 'react';
+import { useDispatch } from 'react-redux';
+import { requestProductsList } from 'calypso/state/products-list/actions';
+import { isProductsListFetching, getProductsListType } from 'calypso/state/products-list/selectors';
+import type { CalypsoDispatch } from 'calypso/state/types';
+import type { AppState } from 'calypso/types';
+
+type QueryProductsListProps = {
+	type?: string;
+	currency?: string;
+	persist?: boolean;
+	productSlugList?: string[];
+};
+
+type RequestProps = Omit< QueryProductsListProps, 'productSlugList' > & {
+	product_slugs?: string[];
+};
+
+const request =
+	( { persist, ...props }: RequestProps ) =>
+	( dispatch: CalypsoDispatch, getState: () => AppState ) => {
+		if (
+			isProductsListFetching( getState() ) ||
+			( persist && props.type === getProductsListType( getState() ) )
+		) {
+			return;
+		}
+
+		dispatch( requestProductsList( props as any ) );
```

Comment:
The cast to `any` should be easy to eliminate: just ensure that `type` is defaulted to `"all"`.